### PR TITLE
Extend existing tests for sales 

### PR DESCRIPTION
**Pull request by szczecha**
I want to merge this change because some shops may be using the old sales and we want to ensure that the tests are more accurate to what the customers are doing.
Changes:
- add more products on sales to checkout
- add products not on sale to checkout
- use line update mutations instead of creating lines only in checkoutCreate
- split parametrized tests for combining sales and vouchers 

to do:
- [X] CORE_1004 [Should be able to buy products on percentage sale in checkout](https://saleor.testmo.net/repositories/5?group_id=132&case_id=2082)
- [x] CORE_1002 [Should be able to buy products on fixed sale in checkout](https://saleor.testmo.net/repositories/5?group_id=132&case_id=2080)
- [x] CORE_0114 [Checkout should calculate correct discount for sale and voucher](https://saleor.testmo.net/repositories/5?group_id=125&case_id=2364)

# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_fixed_sale.py [hunk 1]
```diff
@@ -42,14 +45,42 @@ def prepare_sale_for_product(
...
 
-    return sale_id, sale_discount_value
+    return sale_id
+
+
+def prepare_products(
+    e2e_staff_api_client,
+    warehouse_id,
+    channel_id,
+    number_of_products,
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_fixed_sale.py:62]: `number_of_products` is not needed, you can use `len(prices)` instead

File: saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_fixed_sale.py [hunk 2]
```diff
@@ -109,46 +167,219 @@ def test_checkout_products_on_fixed_sale_core_1002(
...
+    assert order_line1["unitPrice"]["gross"]["amount"] == line1_unit_price
+    assert order_line1["unitDiscount"]["amount"] == fixed_sale_discount_value
+    assert order_line1["unitDiscountReason"] == f"Sale: {sale_id}"
+    assert (
+        order_line1["undiscountedUnitPrice"]["gross"]["amount"]
+        == product2_variant_price
+    )
+
+    assert order_line2["quantity"] == product3_new_quantity
+    assert order_line2["unitPrice"]["gross"]["amount"] == line2_unit_price
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_fixed_sale.py:366]: Should we also check the lines total prices?

File: saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_percentage_sale.py [hunk 3]
```diff
@@ -52,6 +55,34 @@ def prepare_sale_for_product(
     return sale_id, sale_discount_value
 
 
+def prepare_products(
+    e2e_staff_api_client,
+    warehouse_id,
+    channel_id,
+    number_of_products,
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_percentage_sale.py:62]: Same here, `number_of_products` not needed.

File: saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_percentage_sale.py [hunk 4]
```diff
@@ -113,47 +166,225 @@ def test_checkout_products_on_percentage_sale_core_1004(
...
-    line_discount = round(float(product_variant_price) * sale_discount_value / 100, 2)
-    unit_price = float(product_variant_price) - line_discount
+    checkout_line1 = checkout_data["lines"][0]
+    checkout_line2 = checkout_data["lines"][1]
+
+    assert checkout_line1["variant"]["id"] == product1_variant_id
+    assert checkout_line1["quantity"] == product1_quantity
+    line1_discount = round(product1_variant_price * sale_discount_value / 100, 2)
+    line1_unit_price = product1_variant_price - line1_discount
+    assert checkout_line1["unitPrice"]["gross"]["amount"] == line1_unit_price
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_percentage_sale.py:176]: I think we should also check line total prices, in all places, where the `unitPrice` is checked

File: saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_fixed_sale.py [hunk 5]
```diff
@@ -109,46 +167,219 @@ def test_checkout_products_on_fixed_sale_core_1002(
...
+    assert order_line3["unitPrice"]["gross"]["amount"] == line3_unit_price
+    assert order_line3["unitDiscount"]["amount"] == 0
+    assert order_line3["unitDiscountReason"] is None
+    assert (
+        order_line3["undiscountedUnitPrice"]["gross"]["amount"]
+        == product4_variant_price
+    )
+
+    assert order_data["total"]["gross"]["amount"] == calculated_total
+    assert order_data["subtotal"]["gross"]["amount"] == calculated_subtotal
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_fixed_sale.py:384]: Maybe we should also check if the sum of the returned line total prices is equal to `order_data["subtotal"]["gross"]["amount"]`

File: saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_percentage_sale.py [hunk 6]
```diff
@@ -113,47 +166,225 @@ def test_checkout_products_on_percentage_sale_core_1004(
...
+    assert order_line3["unitPrice"]["gross"]["amount"] == line3_unit_price
+    assert order_line3["unitDiscount"]["amount"] == 0
+    assert order_line3["unitDiscountReason"] is None
+    assert (
+        order_line3["undiscountedUnitPrice"]["gross"]["amount"]
+        == product4_variant_price
+    )
+
+    assert order_data["total"]["gross"]["amount"] == calculated_total
+    assert order_data["subtotal"]["gross"]["amount"] == calculated_subtotal
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/sales/test_checkout_product_on_percentage_sale.py:389]: Maybe we should also check if the sum of the returned line total prices is equal to `order_data["subtotal"]["gross"]["amount"]`

File: saleor/tests/e2e/checkout/discounts/test_checkout_calculate_correct_discount_for_sale_and_voucher.py [hunk 7]
```diff
@@ -13,17 +17,48 @@
...
+    checkout_lines_update,
 )
 
 
-def prepare_sale_for_variant(
+def prepare_products(
+    e2e_staff_api_client,
+    warehouse_id,
+    channel_id,
+    number_of_products,
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/test_checkout_calculate_correct_discount_for_sale_and_voucher.py:29]: Same here.

Also, it looks like we are using this method multiple times, maybe it should be separated as some helper and reused?

File: saleor/tests/e2e/checkout/discounts/test_checkout_calculate_correct_discount_for_sale_and_voucher.py [hunk 8]
```diff
@@ -128,40 +137,448 @@ def test_checkout_calculate_discount_for_sale_and_voucher_1014(
...
+        total_gross_amount,
+    )
+
+    # Step 8 - Complete checkout.
+    order_data = checkout_complete(
+        e2e_not_logged_api_client,
+        checkout_id,
+    )
+    assert order_data["status"] == "UNFULFILLED"
+    assert order_data["total"]["gross"]["amount"] == calculated_total
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/test_checkout_calculate_correct_discount_for_sale_and_voucher.py:449]: I think we should also check if the subtotal is equal to the sum of line totals

File: saleor/tests/e2e/checkout/discounts/test_checkout_calculate_correct_discount_for_sale_and_voucher.py [hunk 9]
```diff
@@ -128,40 +137,448 @@ def test_checkout_calculate_discount_for_sale_and_voucher_1014(
...
+
+    assert order_line1["quantity"] == product2_new_quantity
+    assert order_line1["unitDiscountReason"] == (
+        f"Entire order voucher code: {voucher_code} & Sale: {sale_id}"
+    )
+    assert (
+        order_line1["undiscountedUnitPrice"]["gross"]["amount"]
+        == product2_variant_price
+    )
+    assert order_line1["totalPrice"]["gross"]["amount"] == calculated_line1_total
```
Comment (IKarbowiak) [inline saleor/tests/e2e/checkout/discounts/test_checkout_calculate_correct_discount_for_sale_and_voucher.py:470]: Shouldn't we also check the lines unit prices?

**Review by fowczarek — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

**Pull request merged**